scripts/usb_reset.py

```diff
@@ -14,46 +14,266 @@
+#
+# cleanup
+# ./usb_reset.py cleanup -d dom-id
+# ./usb_reset.py cleanup -d 12
+# 1.remove the cgroup if one has been created.
+# 2.umount /dev, /sys from chroot directory if they are mounted.
 
+import argparse
+import ctypes
+import ctypes.util
+import errno
 import fcntl
+import grp
 import xcp.logger as log
 import logging
+import os
+import pwd
 import re
-import sys
-import traceback
+from stat import S_ISCHR, S_ISBLK
+
+
+def parse_arg():
+    parser = argparse.ArgumentParser(
+        description="script to attach, detach, cleanup for USB pass-through")
+    subparsers = parser.add_subparsers(help="command", dest="command")
+
+    attach = subparsers.add_parser("attach", help="attach a USB device")
+    attach.add_argument("device", help="the target usb device")
+    attach.add_argument("-d", dest="dom_id", type=int, required=True,
+                        help="specify the dom_id of the VM")
+    attach.add_argument("-p", dest="pid", type=int, required=True,
+                        help="the process id of QEMU")
+    attach.add_argument("-r", dest="reset_only", type=int,
+                        help="reset device only, for privileged mode")
+
+    detach = subparsers.add_parser("detach", help="detach a USB device")
+    detach.add_argument("device", help="the target usb device")
+    detach.add_argument("-d", dest="dom_id", type=int, required=True,
+                        help="specify the dom_id of the VM")
+    detach.add_argument("-i", dest="id", type=int, nargs=2, required=True,
+                        help="restore ownership to uid gid")
+
+    cleanup = subparsers.add_parser("cleanup", help="clean up chroot directory")
+    cleanup.add_argument("-d", dest="dom_id", type=int, required=True,
+                         help="specify the dom_id of the VM")
+
+    return parser.parse_args()
 
-#USBDEVFS_RESET _IO('U', 20)
-USBDEVFS_RESET = (ord('U') << 8) | 20
 
+def get_root_dir(dom_id):
+    return "/var/xen/qemu/{}/".format(dom_id)
 
+
+def get_cg_dir(dom_id):
+    return "/sys/fs/cgroup/devices/qemu-{}/".format(dom_id)
+
+
+# throw IOError, ValueError
 def read_int(path):
     with open(path) as f:
         return int(f.readline())
 
 
-def reset_device(bus, dev):
-    with open("/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev), "w") as f:
-        fcntl.ioctl(f.fileno(), USBDEVFS_RESET, 0)
+def dev_path(device):
+    pat = re.compile(r"^\d+-\d+(\.\d+)*$")
+    if pat.match(device) is None:
+        log.debug("Unexpected device node: {}".format(device))
+        exit(1)
+    try:
+        bus = read_int("/sys/bus/usb/devices/{}/busnum".format(device))
+        dev = read_int("/sys/bus/usb/devices/{}/devnum".format(device))
+        return "/dev/bus/usb/{0:03d}/{1:03d}".format(bus, dev)
+    except (IOError, ValueError) as e:
+        log.debug("Failed to get device path {}: {}".format(device, str(e)))
+        exit(1)
 
 
-if __name__ == "__main__":
-    log.logToSyslog(level=logging.DEBUG)
+def get_ctl(path, mode):
```

Comment:
What is `ctl`? You could add a docstring.

Reply:
Docstring is still missing

Reply:
done.